qemu/tests/vsock_test.py

```diff
@@ -0,0 +1,149 @@
+    :param session: guest shell session
+    :return: the shell session with opened vsock listening process
+    """
+    nc_vsock_cmd = "%s -l %s" % (nc_vsock_bin, port)
+    session.read_nonblocking(0, timeout=10)
+    logging.info("Listening to the vsock port from guest: %s" % nc_vsock_cmd)
+    session.sendline(nc_vsock_cmd)
+
+
+def nc_vsock_connect(nc_vsock_bin, guest_cid, port):
+    """
+    Connect to vsock port from host, cmd: nc-vsock $guest_cid $port
+
+    :param nc_vsock_bin: path of binary nc-vsock
+    :param guest_cid: guest cid to connect
+    :param port: port to connect
+    :return: The vsock session from host side, being waiting for input
+    """
+    nc_vsock_cmd = "%s %s %s" % (nc_vsock_bin, guest_cid, port)
+    logging.info("Connect to the vsock port on host: %s" % nc_vsock_cmd)
+    return aexpect.Expect(
+        nc_vsock_cmd,
+        auto_close=False,
+        output_func=utils_misc.log_line,
+        output_params=("vsock_%s_%s" % (guest_cid, port),))
+
+
+@error_context.context_aware
+def run(test, params, env):
+    """
+    Vsock basic function test
+
+    1. Boot guest with vhost-vsock-pci device
+    2. Download and compile nc-vsock on both guest and host
+    3. Start listening inside guest, nc-vsock -l $port
+    4. Connect guest CID from host, nc-vsock $guest_cid $port
+    5. Input character, e.g. 'Hello world'
+    6. Check if guest receive the content correctly
+
+    :param test: QEMU test object
+    :param params: Dictionary with the test parameters
+    :param env: Dictionary with test environment
+    """
+
+    def clean():
+        """ Clean the environment """
+        cmd_rm = 'rm -f %s; rm -rf %s*' % (
+            tmp_file, nc_vsock_bin)
+        session.cmd(cmd_rm)
+        process.system(cmd_rm, ignore_status=True)
+        if host_vsock_session.is_alive():
+            host_vsock_session.close()
+        session.close()
+
+    vm = env.get_vm(params["main_vm"])
+    session = vm.wait_for_login()
+    nc_vsock_bin = compile_nc_vsock(vm, session)
+    if nc_vsock_bin is None:
+        clean()
+        test.error("Compile nc-vsock failed")
+    vm_ip_addr = vm.wait_for_get_address(0)
+    port = utils_misc.find_free_port(5000, 6000, vm_ip_addr)
+    vsock_dev = params["vsocks"].split()[0]
+    guest_cid = vm.devices.get(vsock_dev).get_param("guest-cid")
+    nc_vsock_listen(nc_vsock_bin, port, session)
+    host_vsock_session = nc_vsock_connect(nc_vsock_bin, guest_cid, port)
+    connected_str = r"Connection from cid*"
+    send_data = "Hello world"
+    if not session.read_until_last_line_matches(connected_str):
+        clean()
+        test.fail("Connection failed.")
+    error_context.context('Input "Hello world" to vsock.', logging.info)
+    host_vsock_session.sendline(send_data)
+    if not session.read_until_last_line_matches(send_data):
+        clean()
+        test.fail("Data transfer failed.")
+    host_vsock_session.close()
+    session.close()
```

Comment:
@yanan-fu Good question, this is actually two independent test, so I want a fresh clean session for later on test. Actually I am not sure if the guest listening process will exit immediately after the host connection is closed, I am not very confident about the environment after these series of operations.

Reply:
@vivianQizhu host_vsock_session.close() is enough(no need session.close()).  If guest session cann't be closed automatically after host session is closed. it is product bug.   This is one of check point.

Reply:
@xiangchunfu `session.close()` is to close the shell session(where we open nc-vsock process, it will not exit automatically after nc-vsock exit), not the `nc-vsock` process. And there is no such check point in your test case description, would you like to add it in your test case?

Reply:
@vivianQizhu  the result I expected is : nc-vsock process exit automatically inside guest  after nc-vsock process is broken by ctlr+c.  If shell session are closed directly both host and guest.  cann't cover this scenario.  But it is really a extra check point.   If possible, I will add a step(ctrc+c) to porlarion case.

Reply:
@xiangchunfu I think you are talking about another test case, the negative test, this one is for that basic test case(pls check bz), the negative one will be developed with an individual script, not here. Thanks.